### ISSUE TITLE
go/types, types2: type inference should report a better error message

### ISSUE THREAD

**Issue by eliben** [golang/go#48619]
### What version of Go are you using (`go version`)?

<pre>
$  gotip version
go version devel go1.18-fe8347b Fri Sep 24 10:51:48 2021 +0000 linux/amd64
</pre>

### What did you do?

Trying to compile this sample: https://go2goplay.golang.org/p/MAplKhfhuKs

### What did you expect to see?

The sample is incomplete (I tried to minimize it), so I don't expect it to compile. At the very least, there are calls to functions that don't exist.

The code is also incorrect. In the signature of `quickSort_gen_func`, `a, b` should be `int`, not `Elem`. Fixing this, the compiler produces an expected error message instead of going into infinite recursion.

### What did you see instead?

```
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0xc0205e0318 stack=[0xc0205e0000, 0xc0405e0000]
fatal error: stack overflow

runtime stack:
runtime.throw(0x6aa2fc, 0xe)
	/usr/local/go-faketime/src/runtime/panic.go:1126 +0x74
runtime.newstack()
	/usr/local/go-faketime/src/runtime/stack.go:1069 +0x7de
runtime.morestack()
	/usr/local/go-faketime/src/runtime/asm_amd64.s:436 +0x8b

goroutine 1 [running]:
go/types.(*unifier).nifyEq(0xc00058a140, 0x6f32a0, 0xc0000105a0, 0x6f30e8, 0x817000, 0x0, 0x0)
	/usr/local/go-faketime/src/go/types/unify.go:197 +0xb8 fp=0xc0205e0328 sp=0xc0205e0320 pc=0x5a6c18
go/types.(*unifier).nify(0xc00058a140, 0x6f32a0, 0xc0000105a0, 0x6f30e8, 0x817000, 0x0, 0x0)
	/usr/local/go-faketime/src/go/types/unify.go:238 +0x10f6 fp=0xc0205e0488 sp=0xc0205e0328 pc=0x5a7d16
go/types.(*unifier).nifyEq(0xc00058a140, 0x6f32a0, 0xc0000105a0, 0x6f30e8, 0x817000, 0x0, 0x0)
	/usr/local/go-faketime/src/go/types/unify.go:198 +0x70 fp=0xc0205e04d0 sp=0xc0205e0488 pc=0x5a6bd0
go/types.(*unifier).nify(0xc00058a140, 0x6f32a0, 0xc0000105a0, 0x6f30e8, 0x817000, 0x0, 0x0)
	/usr/local/go-faketime/src/go/types/unify.go:238 +0x10f6 fp=0xc0205e0630 sp=0xc0205e04d0 pc=0x5a7d16
...
<SNIP>
```

**Comment by griesemer** [comment]
Smaller reproducer:
```Go
package p

func f[P any](a, _ P) {
	var x int
	f(a, x)
}
```
Compiling leads to an infinite recursion in type inference.

**Comment by cherrymui** [comment]
Any update on this? Thanks.
(checking in as this is a release blocker.)

**Comment by findleyr** [comment]
I de-duped #50273 to this issue. Once this is fixed we should confirm that it also resolves the repro there.

**Comment by griesemer** [comment]
With the above CL this doesn't crash anymore. Still needs a proper fix but not a release blocker anymore.

**Comment by griesemer** [comment]
This produces an ok error message and doesn't crash. Leaving for 1.19.

**Comment by griesemer** [comment]
The type inference algorithm doesn't run into endless recursions anymore since 1.18, and for 1.19 we have disabled the catch-all in case of an infinite recursion and haven't seen any problems so far.

For the simple reproducer
```Go
package p

func f[P any](a, _ P) {
	f(a, int(0))
}
```
the error message is:
```
testdata/manual.go:4:7: type int of int(0) does not match inferred type P for P
```
which is exactly right (barring more information, the 2nd argument's type must be `P`) , and at the correct location.

There's nothing else to do here for now. Closing.
